Design.md

```diff
@@ -0,0 +1,107 @@
+
+As a data backend it must constructs its own metadata in form of a key/string, returned to CloudServer on PUT. This metadata is stored by CloudServer inside its own MongoDB and provided to to the client on GET or DELETE.
+
+All fragments are IMMUTABLE. Any mutation is handled at the CloudServer level.
+
+## Key generation scheme
+
+List of detailed requirements:
+Requirement | Description
+----------- | -------------
+Version | A perfect solution often does not exists. And will not stay that way. Generated keys MUST start with a version, so that we can easily change the format.
+Relate easily to parent CloudServer objectkey | Would make life easy to be able to relate a fragment key with its parent, main object.
+Small and self describing | Smallest keys possible, obviously. But without relying too much on implicit information. Generated keys should be as self describing as possible.
+Embed replication policy | Describe what code is used. Typically RS2+1 or RSk+m or even replication factor for very small object (if we decide to go that way). We could then add new code types easily.
+Split ready | A splitted key should be obvious to spot. The generated key length should NOT depend on number of chunk.
+Fragment name | Resulting fragment key stored on hyperdrive must contain the parent, S3 object name. Or at least a prefix if too long. Resulting key must also be unique hyperdrive-wise.
+Fragment location | Each fragment must enable the hdclient to select the correct hyperdrive
+Fragment type |Each fragment should be tagged as data or coding (reading would then first prefer data parts since RS is systematic)
+
+Since we can specify which key to use on an hyperdrive, we could define a new key scheme to encode fragments in a unified key scheme. The idea is to minimize the amount of metadata we have to store with every object in the MongoDB database that can be used to locate the fragments. Note that we plan to use a 2+1 or 4+2 erasure code for now so the overhead of storing the full uuids for fragment location is not yet important. All fragment IDs should start with the same key as the object (part) key. If the object (part) key is indexed we can quickly find to which object belong the missing fragment (e.g. for repair, see below). The hdclient backend in cloud server will return this constructed key to cloud server main code.
+
+Proposed key generation scheme:
+* <genkey> := <version>#<topology_version>#<split>#<rep_policy>#<object_key>#<rand>#<location>[#<location>]
+* <version> := Natural (so 0 or 1 to start)
+* <topology_version> := Natural (so 0 or 1 to start) - if we use indirection instead of storing raw UUIDS we must versioned the mapping
+* <split> := <n_chunk>,<split_size>
+* <n_chunk> := number of splitted parts (1 for non-splitted objects)
+* <split_size> := size of each splitted parts, except last one (see hyperdrive keys below)
+* <rep_policy> := RS,k,m or CP,n (for n-1 copies)
```

Comment:
What do `RS` and `CP` stand for ? I haven't seen any description yet, though I guess it relates to the algorithm used for storing the actual bits. between various HyperDrive disks ?

Reply:
Will make it more explicit. RS = Reed-Solomon, CP = copy/replication